extensions/sql/processors/ExecuteSQL.cpp

```diff
@@ -0,0 +1,139 @@
+#include "data/JSONSQLWriter.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace processors {
+
+  const std::string ExecuteSQL::ProcessorName("ExecuteSQL");
+
+static core::Property DBCControllerService(
+    core::PropertyBuilder::createProperty("DB Controller Service")->isRequired(true)->withDescription("Database Controller Service.")->supportsExpressionLanguage(true)->build());
+
+static core::Property s_SQLSelectQuery(
+    core::PropertyBuilder::createProperty("SQL select query")->isRequired(true)->withDefaultValue("System")->withDescription(
+        "The SQL select query to execute. The query can be empty, a constant value, or built from attributes using Expression Language. "
+        "If this property is specified, it will be used regardless of the content of incoming flowfiles. "
+        "If this property is empty, the content of the incoming flow file is expected to contain a valid SQL select query, to be issued by the processor to the database. "
+        "Note that Expression Language is not evaluated for flow file contents.")->supportsExpressionLanguage(true)->build());
+
+static core::Property MaxRowsPerFlowFile(
+	core::PropertyBuilder::createProperty("Max Rows Per Flow File")->isRequired(true)->withDefaultValue<int>(0)->withDescription(
+		"The maximum number of result rows that will be included intoi a flow file. If zero then all will be placed into the flow file")->supportsExpressionLanguage(true)->build());
+
+core::Relationship ExecuteSQL::Success("success", "Relationship for successfully consumed events.");
+
+ExecuteSQL::ExecuteSQL(const std::string& name, utils::Identifier uuid)
+    : core::Processor(name, uuid), max_rows_(0),
+      logger_(logging::LoggerFactory<ExecuteSQL>::getLogger()) {
+}
+
+ExecuteSQL::~ExecuteSQL() {
+}
+
+void ExecuteSQL::initialize() {
+  //! Set the supported properties
+  setSupportedProperties( { DBCControllerService, s_SQLSelectQuery, MaxRowsPerFlowFile });
+
+  //! Set the supported relationships
+  setSupportedRelationships( { Success });
+}
+
+void ExecuteSQL::onSchedule(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSessionFactory> &sessionFactory) {
+  context->getProperty(DBCControllerService.getName(), db_controller_service_);
+  context->getProperty(s_SQLSelectQuery.getName(), sqlSelectQuery_);
+  context->getProperty(MaxRowsPerFlowFile.getName(), max_rows_);
+
+  database_service_ = std::dynamic_pointer_cast<sql::controllers::DatabaseService>(context->getControllerService(db_controller_service_));
+  if (database_service_ == nullptr) {
+    throw minifi::Exception(PROCESS_SCHEDULE_EXCEPTION, "DB Controller Service must be defined");
+  }
+}
+
+void ExecuteSQL::onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+  if (database_service_) {
+    std::unique_ptr<sql::Connection> connection = database_service_->getConnection();
```

Comment:
1. Agree, we should reuse connections.

2. Good point, and using mutex makes seance since we don't want to send the same data when QueryDatabaseTable is implemented.

Reply:
Fixed.